pkg/tfbridge/validate_input_types.go

```diff
@@ -0,0 +1,471 @@
+	requiredProps []string,
+) {
+
+	if spec.AdditionalProperties != nil {
+		if spec.AdditionalProperties.Ref != "" {
+			complexSpec = v.getType(spec.AdditionalProperties.Ref)
+			// if we don't find the reference then just continue instead of failing
+			if complexSpec != nil {
+				requiredProps = complexSpec.Required
+			}
+		} else if len(spec.AdditionalProperties.OneOf) > 0 {
+			typeSpec, _, _ = v.findOneOf(propertyValue, spec.OneOf)
+		} else {
+			typeSpec = spec.AdditionalProperties
+		}
+	} else if spec.Ref != "" {
+		complexSpec = v.getType(spec.Ref)
+		if complexSpec != nil {
+			requiredProps = complexSpec.Required
+		}
+	} else if len(spec.OneOf) > 0 {
+		typeSpec, _, _ = v.findOneOf(propertyValue, spec.OneOf)
+		// If the OneOf that we found is an object with a `Type` then return that sub type
+		// Otherwise we should just return the OneOf type that we found
+		if typeSpec.AdditionalProperties != nil && typeSpec.AdditionalProperties.Type != "" {
+			typeSpec = typeSpec.AdditionalProperties
+		}
+	}
+	return complexSpec, typeSpec, requiredProps
+}
+
+// validateTypeSpec is the main function for validating a PropertyValue against the pulumi schema.
+// It is a recursive function that will validate nested types and arrays. It will return a list of
+// type failures if any are found
+func (v *PulumiInputValidator) validateTypeSpec(
+	propertyKey string,
+	propertyValue resource.PropertyValue,
+	typeSpec pschema.TypeSpec,
+	pathBuilder pathBuilder,
+	requiredProps []string,
+) *[]TypeFailure {
+	// for now don't validate discriminators
+	if typeSpec.Discriminator != nil {
+		return nil
+	}
+	// don't type check
+	// - resource references
+	// - assets
+	// - archives
+	if propertyValue.IsResourceReference() || propertyValue.IsAsset() || propertyValue.IsArchive() {
+		return nil
+	}
+
+	// handle non-required properties that are null
+	if propertyValue.TypeString() == "null" {
+		if slices.Contains(requiredProps, propertyKey) {
+			return &[]TypeFailure{
+				{
+					ResourcePath: pathBuilder.toPath(),
+					Reason:       fmt.Sprintf("property %s is required", propertyKey),
+				},
+			}
+		}
+	}
+
+	failures := []TypeFailure{}
+	pathBuilder.add(propertyKey)
+
+	propertyValueType := propertyValue.TypeString()
+	if propertyValue.IsOutput() || propertyValue.IsComputed() || propertyValue.IsSecret() {
+		value := propertyValue.TypeString()
+		// regex should match `secret<string>` or `output<string>`
+		r := regexp.MustCompile(`(?:secret<|output<)+(\w+)>+`)
+		matches := r.FindAllStringSubmatch(value, -1)
+		propertyValueType = matches[0][1]
+	}
+
+	// perform some initial easy type matching. fail fast.
+	possible, ok := v.matchType(propertyValueType, typeSpec)
+	if !ok {
+		return &[]TypeFailure{
+			{
+				ResourcePath: pathBuilder.toPath(),
+				Reason: fmt.Sprintf(
+					"expected %s type, got %s type",
+					possible.toString(),
+					propertyValue.TypeString(),
+				),
+			},
+		}
+	}
+
+	switch propertyValueType {
```

Comment:
Matching based on `propertyValue.TypeString()` is not ideal since that feels like the code becomes sensitive to how these values are presented to the user for debugging. If you unpack how TypeString itself is implemented it's doing matching over its node kind, I think that'd read more naturally here as well:

https://github.com/pulumi/pulumi/blob/master/sdk/go/common/resource/properties.go#L549

Reply:
Sure I can switch to doing it that way.

Reply:
Done!